packages/smooth_app/lib/pages/offline_data.dart

```diff
@@ -0,0 +1,183 @@
+                      title: const Text(
+                        'Offline Product Data',
+                      ),
+                      subtitle: FutureBuilder<int>(
+                        future: daoProduct.getLength(),
+                        builder: (BuildContext context,
+                            AsyncSnapshot<int> snapshot) {
+                          if (snapshot.hasData) {
+                            return Text(
+                              '${snapshot.data} products available for immediate scanning',
+                            );
+                          } else {
+                            return const Text('Loading...');
+                          }
+                        },
+                      ),
+                      trailing: FutureBuilder<double>(
+                        future: daoProduct.getSize(),
+                        builder: (BuildContext context,
+                            AsyncSnapshot<double> snapshot) {
+                          if (snapshot.hasData) {
+                            return Text(
+                              '${snapshot.data} MB',
+                            );
+                          } else {
+                            return const CircularProgressIndicator();
+                          }
+                        },
+                      ),
+                    ),
+                    UserPreferencesListTile(
+                      trailing: const Icon(Icons.refresh),
+                      onTap: () async {
+                        String? status = await LoadingDialog.run<String>(
+                          context: context,
+                          future: daoProduct.getFreshLocalDataBase(),
+                          title: 'Refreshing \n This may take a while',
+                          dismissible: true,
+                        );
+                        if (status == 'OK') {
+                          status = '$length items refreshed';
+                        }
+                        // TODO(ashaman999): dapproductlist.updateTimestamp();
+                        // ignore: use_build_context_synchronously
+                        ScaffoldMessenger.of(context).showSnackBar(
+                          SnackBar(
+                            content: Text(status!),
+                          ),
+                        );
+                      },
+                      title: const Text(
+                        'Update Offline Data',
+                      ),
+                      subtitle: const Text(
+                        'Update the databse with the latest data from the server',
+                      ),
+                    ),
+                    UserPreferencesListTile(
+                      trailing: const Icon(Icons.delete),
+                      onTap: () async {
+                        final int noOdDeletedProducts =
+                            await daoProduct.clearAll();
+                        await daoProductList.clearAll();
```

Comment:
Do you realize that you're also deleting the history and the scan list?
Anyway, you should delete the list, and then the products, not the other way around.
And to be clean, you should put that in a SQL transaction, where you don't have to worry if something bad happens: either it's a success and everything is deleted, or it fails and nothing at all is deleted.

Reply:
> And to be clean, you should put that in a SQL transaction, where you don't have to worry if something bad happens: either it's a success and everything is deleted, or it fails and nothing at all is deleted.

- Getting confused on how to do this 😿